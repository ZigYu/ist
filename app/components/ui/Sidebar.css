.sidebar {
  position: relative;
  width: 200px;
  height: 100vh;
  background: url('../../static/backgrounds/sidebar.png') no-repeat;
  background-size: cover;
  background-color: var(--color-light);
  background-blend-mode: multiply;
  transition: background 1s;
  z-index: 1;
  box-shadow: 0 0 20px -5px black;
}

.contentWrapper {
  overflow-x: hidden;
  overflow-y: auto;
  height: 100%;
}

.content {
  padding: 20px 30px;
  min-width: 200px;
  box-sizing: border-box;
}

.buttonOpen {
  position: absolute;
  top: 10px;
}

.sidebar.left .buttonOpen {
  right: -35px;
}

.sidebar.right .buttonOpen {
  left: -60px;
}
