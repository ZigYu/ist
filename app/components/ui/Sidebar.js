import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './Sidebar.css';
import Animate, { Appear, createPosed } from '../Animate';
import IconHamburger from '../icons/IconHamburger';
import Button from './Button';
import ButtonClose from './ButtonClose';

const cx = classNames.bind(styles);

export const SidebarBox = createPosed({
  open: {
    width: '300px',
    staggerChildren: 50,
    transition: { ease: [0.37, 0.85, 0.53, 0.77], duration: 500 }
  },
  closed: {
    width: '0px',
    transition: { ease: [0.37, 0.85, 0.53, 0.77], duration: 500 }
  }
});

export default class Sidebar extends Component {
  state = { isOpen: false };

  toggleIsOpen = () => {
    const { isOpen } = this.state;
    this.setState({ isOpen: !isOpen });
  };

  render() {
    const { isOpen } = this.state;
    const { children, position, theme } = this.props;

    const className = cx(theme, {
      sidebar: true,
      left: position === 'left',
      right: position === 'right'
    });

    return (
      <SidebarBox className={className} pose={isOpen ? 'open' : 'closed'}>
        <ButtonOpen
          {...this.props}
          isOpen={isOpen}
          onClick={this.toggleIsOpen}
        />
        <div className={styles.contentWrapper}>
          <div className={styles.content}>
            <ButtonClose
              className={styles.buttonClose}
              isOpen={isOpen}
              onClick={this.toggleIsOpen}
            />
            {children}
          </div>
        </div>
      </SidebarBox>
    );
  }
}

Sidebar.propTypes = {
  position: PropTypes.oneOf(['left', 'right']).isRequired,
  openIcon: PropTypes.element
};

Sidebar.defaultProps = {
  openIcon: <IconHamburger />
};

function ButtonOpen({ onClick, isOpen, openIcon }) {
  return (
    <div className={styles.buttonOpen}>
      <Button isTransparent onClick={onClick}>
        <Animate>
          {isOpen ? null : <Appear key="openIcon">{openIcon}</Appear>}
        </Animate>
      </Button>
    </div>
  );
}

ButtonOpen.propTypes = {
  onClick: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  openIcon: PropTypes.element.isRequired
};
