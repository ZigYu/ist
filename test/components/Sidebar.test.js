import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Sidebar from '../../app/components/ui/Sidebar';

Enzyme.configure({ adapter: new Adapter() });

const props = {
  position: 'left',
  children: <div>test</div>
};
const sidebar = shallow(<Sidebar {...props} />);
const clickButtonOpen = () => {
  sidebar.find('ButtonOpen').simulate('click', { stopPropagation: () => {} });
};

describe('<Sidebar />', () => {
  it('Структура Sidebar по умолчанию', () => {
    expect(sidebar).toMatchSnapshot();
  });

  it('Sidebar откроется после клика на Button', () => {
    clickButtonOpen();
    expect(sidebar.state().isOpen).toEqual(true);
  });

  it('Sidebar закроется после клика на Button', () => {
    clickButtonOpen();
    expect(sidebar.state().isOpen).toEqual(false);
  });
});
