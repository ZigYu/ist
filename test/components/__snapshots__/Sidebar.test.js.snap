// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> Структура Sidebar по умолчанию 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Sidebar
    openIcon={
      <IconHamburger
        className=""
      />
    }
    position="left"
  >
    <div>
      test
    </div>
  </Sidebar>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <ButtonOpen
          isOpen={false}
          onClick={[Function]}
          openIcon={
            <IconHamburger
              className=""
            />
          }
          position="left"
        >
          <div>
            test
          </div>
        </ButtonOpen>,
        <div
          className="contentWrapper"
        >
          <div
            className="content"
          >
            <ButtonClose
              className="buttonClose"
              isOpen={false}
              isTransparent={true}
              onClick={[Function]}
            />
            <div>
              test
            </div>
          </div>
        </div>,
      ],
      "className": "sidebar left",
      "pose": "closed",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div>
            test
          </div>,
          "isOpen": false,
          "onClick": [Function],
          "openIcon": <IconHamburger
            className=""
          />,
          "position": "left",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "test",
          },
          "ref": null,
          "rendered": "test",
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="content"
          >
            <ButtonClose
              className="buttonClose"
              isOpen={false}
              isTransparent={true}
              onClick={[Function]}
            />
            <div>
              test
            </div>
          </div>,
          "className": "contentWrapper",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ButtonClose
                className="buttonClose"
                isOpen={false}
                isTransparent={true}
                onClick={[Function]}
              />,
              <div>
                test
              </div>,
            ],
            "className": "content",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "buttonClose",
                "isOpen": false,
                "isTransparent": true,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "test",
              },
              "ref": null,
              "rendered": "test",
              "type": "div",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <ButtonOpen
            isOpen={false}
            onClick={[Function]}
            openIcon={
              <IconHamburger
                className=""
              />
            }
            position="left"
          >
            <div>
              test
            </div>
          </ButtonOpen>,
          <div
            className="contentWrapper"
          >
            <div
              className="content"
            >
              <ButtonClose
                className="buttonClose"
                isOpen={false}
                isTransparent={true}
                onClick={[Function]}
              />
              <div>
                test
              </div>
            </div>
          </div>,
        ],
        "className": "sidebar left",
        "pose": "closed",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <div>
              test
            </div>,
            "isOpen": false,
            "onClick": [Function],
            "openIcon": <IconHamburger
              className=""
            />,
            "position": "left",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "test",
            },
            "ref": null,
            "rendered": "test",
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="content"
            >
              <ButtonClose
                className="buttonClose"
                isOpen={false}
                isTransparent={true}
                onClick={[Function]}
              />
              <div>
                test
              </div>
            </div>,
            "className": "contentWrapper",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <ButtonClose
                  className="buttonClose"
                  isOpen={false}
                  isTransparent={true}
                  onClick={[Function]}
                />,
                <div>
                  test
                </div>,
              ],
              "className": "content",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "buttonClose",
                  "isOpen": false,
                  "isTransparent": true,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "test",
                },
                "ref": null,
                "rendered": "test",
                "type": "div",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
